devlib/module/cgroups2.py

```diff
@@ -0,0 +1,1972 @@
+        """
+        Returns the relative sub-path the delegated root of the V2 hierarchy is mounted on, via the parsing
+        of the /proc/<PID>/cgroup file of the delegated process associated with ``delegated_pid``.
+
+        :param delegated_pid: The Main PID of the transient service unit we requested delegation for.
+        :type delegated_pid: int
+
+        :param target: Interface to target device.
+        :type target: Target
+
+        :return: The sub-path to the delegate root of the V2 CGroup hierarchy.
+        :rtype: str
+        """
+
+        relative_delegated_mount_paths = _read_lines(
+            target=target, path="/proc/{pid}/cgroup".format(pid=delegated_pid)
+        )
+
+        # Following Regex matches the line that contains the relative sub path.
+        REL_PATH_REGEX = re.compile(r"0::\/(?P<path>.+)")
+
+        for mount_path in relative_delegated_mount_paths:
+            m = REL_PATH_REGEX.match(mount_path)
+            if m:
+                return m.group("path")
+            else:
+                raise TargetStableError(
+                    "A V2 CGroup hierarchy was not delegated by systemd."
+                )
+
+    @classmethod
+    def _get_available_controllers(
+        cls, controllers: Dict[str, Dict[str, Union[str, int]]]
+    ):
+        """
+        Returns the CGroup controllers that are currently not in use on the target device,
+        which can be taken control over and used in a manually mounted V2 hierarchy.
+        This method will only be called in the absence of systemd.
+
+        :param controllers: A dictionary of CGroup controller name keys to dictionary value mappings,
+            where the secondary dictionary contains a mapping between various CGroup controller configuration keys
+            and their respectively obtained values for the respective CGroup controllers.
+        :rtype: Dict[str, Dict[str,Union[str,int]]]
+
+        :raises TargetStableError: Occurs in the case where a V2 hierarchy is already mounted on the target device.
+            We want to bail out in this case.
+
+        :return: The ``controllers`` Dict filtered to just those controllers which are free/un-used.
+        :rtype: Dict[str, Dict[str, Union[str,int]]]
+        """
+
+        # Filters the controllers dict to entries where the version is == 2.
+        mounted_v2_controllers = {
+            controller
+            for controller, configuration in controllers.items()
+            if (configuration.get("version") == 2)
+        }
+
+        if mounted_v2_controllers:
+            raise TargetStableError(
+                "A V2 CGroup hierarchy is already mounted on the specified target system, therefore unable to mount requested hierarchy"
+            )
+        else:
+            return {
+                controller: configuration
+                for controller, configuration in controllers.items()
+                if configuration.get("version") is None
+            }
```

Comment:
Could you help me understand where I'm going wrong / the logic here?
I'm attempting to test this on an Android emulator (https://github.com/google/android-emulator-container-scripts) that has cgroups v2, without systemd.
```
generic_x86_64_arm64:/ # cat /proc/cgroups
#subsys_name    hierarchy       num_cgroups     enabled
cpuset          3               6               1
cpu             1               5               1
cpuacct         2               195             1
memory          5               3               1
freezer         4               2               1
```
Inspecting `controllers` I get: 
```
{'cpu': {'hierarchy': 1, 'num_cgroups': 5, 'version': 1, 'mount_point': '/dev/cpuctl'},
 'cpuset': {'hierarchy': 3, 'num_cgroups': 6, 'version': 1, 'mount_point': '/dev/cpuset'},
 'memory': {'hierarchy': 5, 'num_cgroups': 3, 'version': 1, 'mount_point': '/dev/memcg'},
 'freezer': {'hierarchy': 4, 'num_cgroups': 2, 'version': 1, 'mount_point': '/dev/freezer'}}
```

And the function returns that no controllers are available since `version` is not `None` for any of them.

Reply:
Hi Marco,

According to the [cgroups Linux manual page](https://man7.org/linux/man-pages/man7/cgroups.7.html) when it comes to the contents of `/proc/cgroups`,  you actually have CGroups V1 Controllers mounted on the target system. I've attached the section that refers to this for your convenience below.

              The fields in this file are, from left to right:

              1. The name of the controller.

              2. The unique ID of the cgroup hierarchy on which this
                 controller is mounted.  If multiple cgroups v1
                 controllers are bound to the same hierarchy, then each
                 will show the same hierarchy ID in this field.  The
                 value in this field will be 0 if:

                   a) the controller is not mounted on a cgroups v1
                      hierarchy;

                   b) the controller is bound to the cgroups v2 single
                      unified hierarchy; or

                   c) the controller is disabled (see below).

              3. The number of control groups in this hierarchy using
                 this controller.

              4. This field contains the value 1 if this controller is
                 enabled, or 0 if it has been disabled (via the
                 cgroup_disable kernel command-line boot parameter).
 
According to the excerpt above, you do indeed have a CGroups V1 hierarchy mounted on your Target device, which explains what you can see when you inspect `controllers`. Given that there are no free controllers, a V2 hierarchy cannot safely be mounted.

Reply:
Hi Marco,

I've just run the emulator to try to troubleshoot your issue. The reason why this is occurring is because there are no _un-used_ CGroup controllers on the Android device. As you can from the contents of the `/proc/cgroups` file, all the CGroup controllers made available by the kernel have already been mounted during the boot up process, and therefore restricted to use from the POV of the end-user. There are simply no free CGroup controllers.  This explains the error message you receive.

Reply:
Thanks for taking the time to see where I was going wrong and the information. I tried this out on a different setup and managed to recreate your example.